ibis/expr/operations.py

```diff
@@ -1472,22 +1472,20 @@ class OuterJoin(Join):
     pass
 
 
-class LeftSemiJoin(Join):
+class InnerSemiJoin(Join):
```

Comment:
How is this different from a `LEFT SEMI JOIN` (or `ANY LEFT JOIN` in clickhouse parlance)?

If I understand the [clickhouse `JOIN` documentation](https://clickhouse.yandex/docs/en/query_language/queries.html#join-clause) correctly it isn't.

* `If INNER is specified, the result will contain only those rows that have a matching row in the right table.`
* `If ANY is specified and there are multiple matching rows in the right table, only the first one will be joined.`

Composing those two statements yields (paraphrasing):

* `If INNER is specified and ANY is specified the result will contain only the first row from the left table where there's a matching row in the right table.`

Adding `LEFT` into the mix:

* `LEFT is specified, any rows in the left table that don’t have matching rows in the right table will be assigned the default value - zeros or empty rows.`

Combining that will `ANY` from above would yield the same statement composed from the definitions of `ANY` and `INNER` that I wrote above.

Is there something I'm missing here?

Reply:
Actually I've got confused, but [here are the test cases](https://github.com/yandex/ClickHouse/tree/1d836b2bf8fef378d47258957ff74ed3a4aff136/dbms/tests/queries/0_stateless) from 49 to 53

Reply:
There are four type of joins so must be a difference.

Reply:
So it turns out there's a difference :)

* `ANY INNER JOIN` is most similar to a `LEFT SEMI JOIN` (which ibis supports)
* `ANY LEFT JOIN` has no equivalent single node operation in ibis.

I say "most similar to" instead of "equivalent" because the test suite shows that you can select rows from the right table.

In every relational database that I've ever used that supports a `LEFT SEMI JOIN` syntax you can only select fields from the left table. Databases that *don't* implement a syntax for this operation implement it through an `EXISTS` query, which precludes selecting from the right table (since there is no right table in that case).

The reason for this difference is probably because the relational algebraic definition of a semijoin only includes tuples from the left table.

This means that to fully support clickhouse joins we need to introduce new types of joins that are only available to the clickhouse backend.

You can implement this kind of join in other databases using `ROW_NUMBER() OVER () AS i` on the left table and filtering out only those rows where `i = 1` (the first row to be joined).

I think the solution here is to bring back the `any_left_join` etc methods and we can implement them in other databases down the line as desired.